File: api/group/group_service.go
package group

import (
	"context"
	"sort"

	"github.com/larscom/gitlab-ci-dashboard/model"

	"github.com/larscom/gitlab-ci-dashboard/config"
)

type GroupService interface {
	GetGroups(ctx context.Context) ([]model.Group, error)
}

type groupService struct {
	config *config.GitlabConfig
	client GroupClient
}

func NewService(config *config.GitlabConfig, client GroupClient) GroupService {
	return &groupService{
		config: config,
		client: client,
	}
}

func (s *groupService) GetGroups(ctx context.Context) ([]model.Group, error) {
	if len(s.config.GroupOnlyIds) > 0 {
		groups, err := s.client.GetGroupsById(s.config.GroupOnlyIds, ctx)
		return sortByName(groups), err
	}
	groups, err := s.client.GetGroups(ctx)
	return sortByName(groups), err
}

func sortByName(groups []model.Group) []model.Group {
	sort.SliceStable(groups[:], func(i, j int) bool {
		return groups[i].Name < groups[j].Name
	})
	return groups
}

File: api/src/group/group_service.rs
use crate::config::Config;
use crate::error::ApiError;
use crate::gitlab::GitlabApi;
use crate::model::Group;
use moka::future::Cache;
use std::sync::Arc;

pub struct GroupService {
    cache: Cache<String, Vec<Group>>,
    client: Arc<dyn GitlabApi>,
    config: Config,
}

impl GroupService {
    pub fn new(client: Arc<dyn GitlabApi>, config: Config) -> Self {
        let cache = Cache::builder()
            .time_to_live(config.ttl_group_cache)
            .build();

        Self {
            cache,
            client,
            config,
        }
    }
}

impl GroupService {
    pub async fn get_groups(&self, cache_key: &str) -> Result<Vec<Group>, ApiError> {
        let only_ids = &self.config.group_only_ids;
        let skip_groups = &self.config.group_skip_ids;
        let top_level = self.config.group_only_top_level;
        self.cache
            .try_get_with_by_ref(cache_key, async {
                let mut groups = self
                    .client
                    .groups(skip_groups, top_level)
                    .await?
                    .into_iter()
                    .filter(|group| only_ids.is_empty() || only_ids.contains(&group.id))
                    .collect::<Vec<_>>();

                groups.sort_unstable_by(|a, b| a.name.cmp(&b.name));

                Ok::<Vec<Group>, ApiError>(groups)
            })
            .await
            .map_err(|error| error.as_ref().to_owned())
    }
}
